File: bindings/java/src/main/java/io/omq/Context.java
package io.omq;

import java.lang.ref.Cleaner;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/** Owns native OMQ I/O threads and creates sockets. */
public final class Context implements AutoCloseable {
    private static final Cleaner CLEANER = Cleaner.create();

    private final Set<Socket.State> sockets = ConcurrentHashMap.newKeySet();
    private final State state;
    private final Cleaner.Cleanable cleanable;

    /** Creates a context with one native I/O thread. */
    public Context() {
        this(1);
    }

    /** Creates a context with the requested native I/O thread count. */
    public Context(int ioThreads) {
        if (ioThreads <= 0) {
            throw new IllegalArgumentException("ioThreads must be greater than zero");
        }
        this.state = new State(Native.contextCreate(ioThreads));
        this.cleanable = CLEANER.register(this, state);
    }

    private Context(long handle, boolean owner) {
        this.state = new State(handle, owner);
        this.cleanable = CLEANER.register(this, state);
    }

    /** Opens a context with one native I/O thread. */
    public static Context open() {
        return new Context();
    }

    /** Opens a context with the requested native I/O thread count. */
    public static Context open(int ioThreads) {
        return new Context(ioThreads);
    }

    /** Imports a context handle by an opaque process-local share key. */
    public static Optional<Context> fromShareKey(UUID key) {
        Objects.requireNonNull(key, "key");
        long handle = Native.contextFromShareKey(
                key.getMostSignificantBits(), key.getLeastSignificantBits());
        if (handle == 0) {
            return Optional.empty();
        }
        return Optional.of(new Context(handle, false));
    }

    /** Returns an opaque process-local key for importing this context. */
    public UUID shareKey() {
        synchronized (state) {
            long[] key = Native.contextShareKey(state.handle());
            return new UUID(key[0], key[1]);
        }
    }

    /** Creates a socket owned by this context. */
    public Socket socket(SocketType type) {
        Objects.requireNonNull(type, "type");
        synchronized (state) {
            return new Socket(this, state.handle(), type, sockets);
        }
    }

    void remove(Socket.State state) {
        sockets.remove(state);
    }

    /** Closes all owned sockets and terminates the native context. */
    @Override
    public void close() {
        synchronized (state) {
            for (Socket.State socket : sockets.toArray(Socket.State[]::new)) {
                socket.close();
            }
            cleanable.clean();
        }
    }

    private static final class State implements Runnable {
        private final AtomicLong handle;
        private final boolean owner;

        private State(long handle) {
            this(handle, true);
        }

        private State(long handle, boolean owner) {
            this.handle = new AtomicLong(handle);
            this.owner = owner;
        }

        @Override
        public synchronized void run() {
            long handle = this.handle.getAndSet(0);
            if (handle != 0) {
                Native.contextClose(handle, owner);
            }
        }

        synchronized long handle() {
            long handle = this.handle.get();
            if (handle == 0) {
                throw new ClosedException("context closed");
            }
            return handle;
        }
    }
}

File: bindings/go/context.go
package omq

import (
	"context"
	"runtime"
	"sync"
	"sync/atomic"
)

// Config controls native context and Go ring defaults.
type Config struct {
	// IOThreads sets native OMQ I/O thread count.
	IOThreads int
	// RingSize sets private Go/native ring descriptor capacity.
	RingSize int
	// OverrunPolicy sets default channel overrun behavior.
	OverrunPolicy OverrunPolicy
}

// Context owns a native OMQ context and its sockets.
type Context struct {
	handle    *nativeContext
	ringSize  int
	overrun   OverrunPolicy
	mu        sync.Mutex
	sockets   map[*Socket]struct{}
	closed    atomic.Bool
	closeOnce sync.Once
	freeOnce  sync.Once
	closeDone chan struct{}
}

// Open creates a native OMQ context.
func Open(config Config) (*Context, error) {
	ioThreads := config.IOThreads
	if ioThreads == 0 {
		ioThreads = 1
	}
	if ioThreads < 0 {
		return nil, &ConfigError{Err: "io_threads must be greater than zero"}
	}
	if config.RingSize < 0 {
		return nil, &ConfigError{Err: "ring size must be non-negative"}
	}
	handle, err := contextOpenNative(ioThreads)
	if err != nil {
		return nil, err
	}
	ctx := &Context{
		handle:    handle,
		ringSize:  config.RingSize,
		overrun:   config.OverrunPolicy,
		sockets:   make(map[*Socket]struct{}),
		closeDone: make(chan struct{}),
	}
	runtime.SetFinalizer(ctx, (*Context).free)
	return ctx, nil
}

// OpenShared imports a process-local shared native context.
func OpenShared(key ShareKey) (*Context, error) {
	handle, err := contextFromShareKeyNative(key)
	if err != nil {
		return nil, err
	}
	ctx := &Context{
		handle:    handle,
		sockets:   make(map[*Socket]struct{}),
		closeDone: make(chan struct{}),
	}
	runtime.SetFinalizer(ctx, (*Context).free)
	return ctx, nil
}

// ShareKey returns a process-local key for sharing this context.
func (c *Context) ShareKey() (ShareKey, error) {
	if c == nil || c.handle == nil || c.closed.Load() {
		return ShareKey{}, ErrClosed
	}
	key, err := contextShareKeyNative(c.handle)
	keepAlive(c)
	return key, err
}

// Socket creates a socket and applies pre-I/O options.
func (c *Context) Socket(socketType SocketType, opts ...SocketOption) (*Socket, error) {
	if c == nil || c.handle == nil || c.closed.Load() {
		return nil, ErrClosed
	}
	c.mu.Lock()
	if c.closed.Load() {
		c.mu.Unlock()
		return nil, ErrClosed
	}
	handle, err := socketNewNative(c.handle, socketType)
	if err != nil {
		c.mu.Unlock()
		return nil, err
	}
	socket := newSocket(handle, socketType, c, c.ringSize, c.overrun)
	runtime.SetFinalizer(socket, (*Socket).free)
	c.sockets[socket] = struct{}{}
	c.mu.Unlock()
	for _, opt := range opts {
		if err := opt(socket); err != nil {
			socket.free()
			return nil, err
		}
	}
	keepAlive(c)
	return socket, nil
}

// Close closes all owned sockets and terminates the context.
func (c *Context) Close() error {
	return c.CloseContext(context.Background())
}

// CloseContext closes the context, bounded by ctx.
func (c *Context) CloseContext(ctx context.Context) error {
	if ctx == nil {
		ctx = context.Background()
	}
	if c == nil || c.handle == nil {
		return nil
	}
	c.closeOnce.Do(func() {
		c.closed.Store(true)
		go c.closeAll()
	})
	select {
	case <-c.closeDone:
		keepAlive(c)
		return nil
	case <-ctx.Done():
		return errFromContext(ctx)
	}
}

func (c *Context) free() {
	if c == nil || c.handle == nil {
		return
	}
	c.closeOnce.Do(func() {
		c.closed.Store(true)
		c.closeAll()
	})
	<-c.closeDone
}

func (c *Context) removeSocket(socket *Socket) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.sockets, socket)
}

func (c *Context) closeAll() {
	defer close(c.closeDone)
	sockets := c.detachSockets()
	for _, socket := range sockets {
		_ = socket.Close(context.Background())
	}
	c.freeOnce.Do(func() {
		contextFreeNative(c.handle)
	})
	runtime.SetFinalizer(c, nil)
}

func (c *Context) detachSockets() []*Socket {
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.sockets) == 0 {
		return nil
	}
	sockets := make([]*Socket, 0, len(c.sockets))
	for socket := range c.sockets {
		sockets = append(sockets, socket)
	}
	c.sockets = make(map[*Socket]struct{})
	return sockets
}
